src/crimson/os/seastore/linked_tree_node.h

```diff
@@ -0,0 +1,973 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "crimson/os/seastore/cached_extent.h"
+#include "crimson/os/seastore/transaction.h"
+
+namespace crimson::os::seastore {
+
+template <typename ParentT>
+class child_pos_t {
+public:
+  child_pos_t(TCachedExtentRef<ParentT> stable_parent, uint16_t pos)
+    : stable_parent(stable_parent), pos(pos) {}
+
+  TCachedExtentRef<ParentT> get_parent() {
+    ceph_assert(stable_parent);
+    return stable_parent;
+  }
+  uint16_t get_pos() {
+    return pos;
+  }
+  template <typename ChildT>
+  void link_child(ChildT *c) {
+    get_parent()->link_child(c, pos);
+  }
+private:
+  TCachedExtentRef<ParentT> stable_parent;
+  uint16_t pos = std::numeric_limits<uint16_t>::max();
+};
+
+using get_child_iertr = trans_iertr<crimson::errorator<
+  crimson::ct_error::input_output_error>>;
+template <typename T>
+using get_child_ifut = get_child_iertr::future<TCachedExtentRef<T>>;
+
+template <typename ParentT, typename ChildT>
+struct get_child_ret_t {
+  std::variant<child_pos_t<ParentT>, get_child_ifut<ChildT>> ret;
+  get_child_ret_t(child_pos_t<ParentT> pos)
+    : ret(std::move(pos)) {}
+  get_child_ret_t(get_child_ifut<ChildT> child)
+    : ret(std::move(child)) {}
+
+  bool has_child() const {
+    return ret.index() == 1;
+  }
+
+  child_pos_t<ParentT> &get_child_pos() {
+    ceph_assert(ret.index() == 0);
+    return std::get<0>(ret);
+  }
+
+  get_child_ifut<ChildT> &get_child_fut() {
+    ceph_assert(ret.index() == 1);
+    return std::get<1>(ret);
+  }
+};
+
+template <typename T, typename key_t>
+struct node_cmp_t {
+  using is_transparent = key_t;
+  bool operator()(const TCachedExtentRef<T> &l,
+		  const TCachedExtentRef<T> &r) const {
+    assert(l->get_end() <= r->get_begin()
+      || r->get_end() <= l->get_begin()
+      || (l->get_begin() == r->get_begin()
+	  && l->get_end() == r->get_end()));
+    return l->get_begin() < r->get_begin();
+  }
+  bool operator()(const key_t &l, const TCachedExtentRef<T> &r) const {
+    return l < r->get_begin();
+  }
+  bool operator()(const TCachedExtentRef<T> &l, const key_t &r) const {
+    return l->get_begin() < r;
+  }
+};
+
+template <
+  typename ParentT,
+  typename node_key_t,
+  typename Comparator = node_cmp_t<ParentT, node_key_t>>
+class ParentNode;
+
+// link the root of the tree with its parent
+template <typename ParentT, typename RootT>
+class TreeRootLinker {
+public:
+  static void link_root(TCachedExtentRef<ParentT> &root_parent, RootT* root_node);
+  static void unlink_root(TCachedExtentRef<ParentT> &root_parent);
+};
+
+// RootNode is the root of the tree.
+template <typename ParentT, typename T>
```

Comment:
Let's comment that `ParentT` must be `RootBlock`, if cannot be asserted or hardcoded here.